## Setup

Run the code below to load the scrambled message:


In [1]:
import pandas as pd

message = pd.read_csv("https://www.dropbox.com/s/lgpn3vmksk3ssdo/scrambled_message.txt?dl=1")['Word']

In [2]:
message

0                    Koila!
1                     In   
2                     kiew,
3                         a
4                 humble   
               ...         
122                     you
123                 mabugh.
124              ughhh?call
125        meugh.ughhhh!   
126                      K.
Name: Word, Length: 127, dtype: object

In this activity, a "word" refers to any set of characters with no white space, even though they are not truly an English word.  That is, even though many of elements of the scrambled message vector are nonsense, and some have punctuation, you can consider each element to be a "word".

Beware!  The object named `message` is a **pandas Series** of strings.  If you want to use functions that expect a string, rather than a series, remember `.apply()` and `lambda` functions.




## Warm-up exercises

1. How many characters are in the scrambled message?
2. How many of these characters are white space?
3. How many words are in the scrambled message?
4. Show all the punctuation marks in the scrambled message.
5. Print out, in all capitals, the longest word in the scrambled message.
6. Print out every piece of a word that starts with the letter "m" and ends with the letter "z" in the scrambled message.

In [22]:
length = message.apply(lambda string: len(string))
sum(length)

2544

There are 2544 characters in the secret message.

In [21]:
lengths = message.apply(lambda string: len(string.strip()))
sum(lengths)

892

In [23]:
2544-892

1652

There are 892 characters in the secret message that are not whitespace, making 1652 characters whitespace.

In [39]:
import re
import numpy as np

In [42]:
words = message.apply(lambda string: string.strip())

words = words.apply(lambda x: re.findall(r'^..+', x))

words = np.concatenate(words)

len(words)


122

There are 122 words in the message.

In [ ]:

punct = message.apply(lambda x: re.findall(r'[^\w\d]', x))




## Decode a message

Complete the following steps to decode the message.  

1. Remove any spaces before or after each word.
2. Any time you see the word "ugh", with any number of h's, followed by a punctuation mark, delete this.
3. No word should be longer than 16 characters. Drop all extra characters beyond 13 off the end of each word.
4. Replace all instances of exactly 2 a's with exactly 2 e's.
5. Replace all z's with t's.
6. Every word that ends in b, change that to a y.  *Hint: look out for punctuation!*
7. Every word that starts with k, change that to a v.  *Hint: look out for capitalization!*
8. Use `.join()` to recombine all your words into a message.
9. Find the movie this quote is from.

In [94]:
words = message.apply(lambda string: string.strip())


words = words.apply(lambda ugh: re.sub(r'ugh+[^\w\s\d]', '', ugh))


In [95]:
words = words.apply(lambda x: x[0:13])

In [96]:
words = words.apply(lambda ugh: re.sub('aa', 'ee', ugh))

In [97]:
words = words.apply(lambda ugh: re.sub('z', 't', ugh))

In [98]:
words = words.apply(lambda ugh: re.sub(r'b$', 'y', ugh))

In [99]:
words = words.apply(lambda ugh: re.sub(r'^[kK]', 'V', ugh))

In [102]:
' '.join(words)

'Voila! In View, a humble Vaudevilliana Veteran, cast Vicariously as both Victim and Villain by the Vicissitudes of fate. This Visage, no mere Veneer of Vanity, is a Vestige of the Vox populi now Vacant, Vanished. However, this Valorous Visitation of a bygone Vexation stands Vivified, and has Vowed to Vanquish these Venal and Virulent Vermin, Van guarding Vice and Vouchsafing the Violently Vicious and Voracious Violation of Volition. The only Verdict is Vengeance; a Vendetta, held as a Votive not in Vain, for the Value and Veracity of such shall one day Vindicate the Vigilant and the Virtuous. Verily this Vichyssoise of Verbiage Veers most Verbose, so let me simply add that its my Very good honour to meet you and you may call me V.'